cou/apps/app.py

```diff
@@ -182,7 +276,279 @@ def _get_os_origin(self) -> str:
         """
         for origin in ("openstack-origin", "source"):
             if self.config.get(origin):
+                self.origin_setting = origin
                 return self.config[origin].get("value", "")
 
         logger.warning("Failed to get origin for %s, no origin config found", self.name)
         return ""
+
+    async def check_upgrade(self) -> None:
+        """Check if an application has upgraded its workload version."""
+        status = await async_get_status()
+        app_status = status.applications.get(self.name)
+        units_not_upgraded = []
+        for unit in app_status.units.keys():
+            workload_version = app_status.units[unit].workload_version
+            compatible_os_versions = OpenStackCodenameLookup.lookup(self.charm, workload_version)
+            if self.next_os_release not in compatible_os_versions:
+                units_not_upgraded.append(unit)
+        if units_not_upgraded:
+            logger.error(
+                "App: '%s' has units: '%s' didn't upgrade to %s",
+                self.name,
+                ", ".join(units_not_upgraded),
+                self.next_os_release,
+            )
+
+    def can_generate_upgrade_plan(self) -> bool:
+        """Check if application can generate an upgrade plan.
+
+        Applications where it's not possible to identify the current OpenStack release,
+        should not generate upgrade plan.
+        :return: True if can generate a plan, else False.
+        :rtype: bool
+        """
+        if not self.current_os_release:
```

Comment:
We can't generate upgrade plan if `self.next_os_release` is None, so I think we should:
```suggestion
        if not self.next_os_release:
```

Reply:
True, and this also indirectly cover the situation when `current_os_release` is None.

Done